src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -5403,20 +5403,16 @@ public Type constantType(LoadableConstant c) {
      * Do t and s have the same nullability?
      */
     public boolean hasSameNullability(Type t, Type s) {
+        // special case for literals, a literal is always != null
+        boolean isLiteral = s != null && s.getMetadata(TypeMetadata.ConstantValue.class) != null;
```

Comment:
I believe this should be addressed at the level of the `isNullUnspecified` predicates. E.g. if something is a constant type, then you know it's null-restricted (e.g. because we infer it so). Once that's dealt with, I believe these higher-level methods here can stay unchanged?

Reply:
yes I did that experiment first but then I got to some unexpected issues. For example: `1L` now would be consider to have type `long!` which doesn't make too much sense, we could exclude primitives from the predicates I guess but there are still some other places where we can have surprises. I found this example:

```
import java.lang.invoke.*;
class Cell {
    final void reset() {
        /* we are testing that the compiler won't infer the arguments of
         * VarHandle::setVolatile as (Cell, String!)
         */
        VALUE.setVolatile(this, "");
    }
    final void reset(String identity) {
        /* if that were the case, see comment above, then this invocation would generate
         * a warning, VarHandle::setVolatile is a polymorphic signature method
         */
        VALUE.setVolatile(this, identity);
    }
                                                                
    private static final VarHandle VALUE;
    static {
        try {
            MethodHandles.Lookup l = MethodHandles.lookup();
            VALUE = l.findVarHandle(Cell.class, "value", long.class);
        } catch (ReflectiveOperationException e) {
            throw new ExceptionInInitializerError(e);
        }
    }
}
```
VarHandle::setVolatile is a polymorphic signature method. So from the first invocation the compiler would infer argument types: `(Cell, String!)` is this what we want? For the second invocation we find that we already have a method that matches but the invocation will generate a warning if compiled with `Xlint:null`, not sure we want this. This is why I preferred to stay with the declared type for literals and tread them as a special case. We can still do what you suggest and do something different when we infer the signature of polymorphic sig methods, like not consider nullability.